samples/triangle/main.c

```diff
@@ -191,59 +190,39 @@ static void init_shader(void)
     p = pb_begin();
 
     /* Set run address of shader */
-    pb_push1(p, NV097_SET_TRANSFORM_PROGRAM_START, 0);
-    p += 2;
+    p = xgu_set_transform_program_start(p, 0);
 
     /* Set execution mode */
-    pb_push1(p, NV097_SET_TRANSFORM_EXECUTION_MODE,
-        MASK(NV097_SET_TRANSFORM_EXECUTION_MODE_MODE, NV097_SET_TRANSFORM_EXECUTION_MODE_MODE_PROGRAM)
-        | MASK(NV097_SET_TRANSFORM_EXECUTION_MODE_RANGE_MODE, NV097_SET_TRANSFORM_EXECUTION_MODE_RANGE_MODE_PRIV));
-    p += 2;
-
-    pb_push1(p, NV097_SET_TRANSFORM_PROGRAM_CXT_WRITE_EN, 0);
-    p += 2;
+    p = xgu_set_transform_execution_mode(p, XGU_PROGRAM, XGU_RANGE_MODE_PRIVATE);
+    p = xgu_set_transform_program_cxt_write_enabled(p, false);
 
     /* Set cursor and begin copying program */
-    pb_push1(p, NV097_SET_TRANSFORM_PROGRAM_LOAD, 0);
-    p += 2;
-
-    for (i=0; i<sizeof(vs_program)/8; i++) {
-        pb_push(p++, NV097_SET_TRANSFORM_PROGRAM, 4);
-        memcpy(p, &vs_program[i*4], 4*4);
-        p+=4;
-    }
+    p = xgu_set_transform_program_load(p, 0);
+    p = xgu_set_transform_program(p, vs_program, sizeof(vs_program)/16);
 
-    pb_end(p);
+    p = pb_end(p);
 
     /* Setup fragment shader */
     p = pb_begin();
     #include "ps.inl"
-    pb_end(p);
+    p = pb_end(p);
 }
 
 /* Set an attribute pointer */
-static void set_attrib_pointer(unsigned int index, unsigned int format, unsigned int size, unsigned int stride, const void* data)
+static void set_attrib_pointer(XguVertexArray index, XguVertexArrayType format, unsigned int size, unsigned int stride, const void* data)
 {
     uint32_t *p = pb_begin();
-    pb_push1(p, NV097_SET_VERTEX_DATA_ARRAY_FORMAT + index*4,
-        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE, format) | \
-        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_SIZE, size) | \
-        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_STRIDE, stride));
-    p += 2;
-    pb_push1(p, NV097_SET_VERTEX_DATA_ARRAY_OFFSET + index*4, (uint32_t)data & 0x03ffffff);
-    p += 2;
-    pb_end(p);
+    p = xgu_set_vertex_data_array_format(p, index, format, size, stride);
+    p = xgu_set_vertex_data_array_offset(p, index, (uint32_t)data & 0x03ffffff);
+    p = pb_end(p);
 }
 
 /* Send draw commands for the triangles */
-static void draw_arrays(unsigned int mode, int start, int count)
+static void draw_arrays(XguPrimitiveType mode, int start, int count)
 {
     uint32_t *p = pb_begin();
-    pb_push1(p, NV097_SET_BEGIN_END, mode); p += 2;
-
-    pb_push(p++,0x40000000|NV097_DRAW_ARRAYS,1); //bit 30 means all params go to same register 0x1810
-    *(p++) = MASK(NV097_DRAW_ARRAYS_COUNT, (count-1)) | MASK(NV097_DRAW_ARRAYS_START_INDEX, start);
-
-    pb_push1(p,NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END); p += 2;
-    pb_end(p);
+    p = xgu_begin(p, mode);
+    p = xgu_draw_arrays(p, start, count - 1); //FIXME: Off-by-one error?
+    p = xgu_end(p);
+    p = pb_end(p);
 }
```

Comment:
This is a good candidate to move to `void xgux_draw_arrays(XguPrimitiveType mode, unsigned int start, unsigned int count)` (where XGUX is a higher-than XGU but lower level than GL API - think GL/GLU)